## Verify Data Quality

### Data manipulation

In this section, I will import the data and make any transforms required for this section. This should be merged into the general data prep for the rest of the notebook

In [15]:
import pandas as pd
import pandas_profiling as pp
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

In [6]:
airbnb = pd.read_csv('C:/Users/William/OneDrive/MLI/AB_NYC_2019.csv')
airbnb.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

###  Explain any missing values

Stuff

In [7]:
airbnb.isnull().sum()

id                                    0
name                                 16
host_id                               0
host_name                            21
neighbourhood_group                   0
neighbourhood                         0
latitude                              0
longitude                             0
room_type                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
last_review                       10052
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

### Duplicate Data

In [9]:
airbnbDups = airbnb[airbnb.duplicated()]
airbnbDups.head()

Empty DataFrame
Columns: [id, name, host_id, host_name, neighbourhood_group, neighbourhood, latitude, longitude, room_type, price, minimum_nights, number_of_reviews, last_review, reviews_per_month, calculated_host_listings_count, availability_365]
Index: []

There are no duplicate rows in the data, based on the check above.

### Outliers

In [17]:
airbnb.drop(columns=['host_id','id']).describe(percentiles=[.01, .05, .25, .5, .75, .95, .99])


latitude     longitude         price  minimum_nights  \
count  48895.000000  48895.000000  48895.000000    48895.000000   
mean      40.728949    -73.952170    152.720687        7.029962   
std        0.054530      0.046157    240.154170       20.510550   
min       40.499790    -74.244420      0.000000        1.000000   
1%        40.596687    -74.026774     30.000000        1.000000   
5%        40.646114    -74.003880     40.000000        1.000000   
25%       40.690100    -73.983070     69.000000        1.000000   
50%       40.723070    -73.955680    106.000000        3.000000   
75%       40.763115    -73.936275    175.000000        5.000000   
95%       40.825643    -73.865771    355.000000       30.000000   
99%       40.864661    -73.776920    799.000000       45.000000   
max       40.913060    -73.712990  10000.000000     1250.000000   

       number_of_reviews  reviews_per_month  calculated_host_listings_count  \
count       48895.000000       38843.000000                    48895.000000   
mean           23.274466           1.373221                        7.143982   
std            44.550582           1.680442                       32.952519   
min             0.000000           0.010000                        1.000000   
1%              0.000000           0.020000                        1.000000   
5%              0.000000           0.040000                        1.000000   
25%             1.000000           0.190000                        1.000000   
50%             5.000000           0.720000                        1.000000   
75%            24.000000           2.020000                        2.000000   
95%           114.000000           4.640000                       15.000000   
99%           214.000000           7.195800                      232.000000   
max           629.000000          58.500000                      327.000000   

       availability_365  
count      48895.000000  
mean         112.781327  
std          131.622289  
min            0.000000  
1%             0.000000  
5%             0.000000  
25%            0.000000  
50%           45.000000  
75%          227.000000  
95%          359.000000  
99%          365.000000  
max          365.000000

Outliers would not be relevant for the id and categorical columns. So id, host_id, name, host_name, and room_type would not need attention. Categorical variables would be checked for unusual categories. 

Other columns, based on the describe function above, seem to have reasonable values. Latitude and Longitude are within a tight range, suggesting we are not seeing listings far away from NYC. availability_365 seems ok for the most part. A large number of listings seem to be available zero nights, these could be a mistake or perhaps this field was not collected for all listings. There are no numbers above 365 either, and it is reasonable for a host to offer their property all year long.

We will take a look at the others on an individual basis. Lets start with price.

In [29]:
airbnb[['name','price','minimum_nights','room_type','neighbourhood']].sort_values(by='price', ascending=False)

name  price  \
9151                 Furnished room in Astoria apartment  10000   
17692    Luxury 1 bedroom apt. -stunning Manhattan views  10000   
29238                                1-BR Lincoln Center  10000   
40433  2br - The Heart of NYC: Manhattans Lower East ...   9999   
12342                Quiet, Clean, Lit @ LES & Chinatown   9999   
6530                                  Spanish Harlem Apt   9999   
30268  Beautiful/Spacious 1 bed luxury flat-TriBeCa/Soho   8500   
4377                                       Film Location   8000   
29662          East 72nd Townhouse by (Hidden by Airbnb)   7703   
42523                70' Luxury MotorYacht on the Hudson   7500   
45666                               Gem of east Flatbush   7500   
44034                   3000 sq ft daylight photo studio   6800   
37194               Apartment New York \nHell’s Kitchens   6500   
3774                    SUPER BOWL Brooklyn Duplex Apt!!   6500   
48043       Luxury TriBeCa Apartment at an amazing price   6500   
29664          Park Avenue Mansion by (Hidden by Airbnb)   6419   
3537                  UWS 1BR w/backyard + block from CP   6000   
15560                 Luxury townhouse Greenwich Village   6000   
3720                 SuperBowl Penthouse Loft 3,000 sqft   5250   
43009    Midtown Manhattan great location (Gramacy park)   5100   
22353                            Victorian Film location   5000   
26739                                         Broadway 1   5000   
38000      4-Floor Unique Event Space 50P Cap. - #10299B   5000   
25825                                           Fulton 2   5000   
2698                 Beautiful 1 Bedroom in Nolita/Soho    5000   
4345                    NearWilliamsburg bridge 11211 BK   5000   
46614                     Northside Williamsburg Stunner   4500   
4376                                Photography Location   4500   
43670   Beautiful private Brooklyn room with kitchenette   4200   
45867        bay ridge & sunset park furnished apartment   4200   
...                                                  ...    ...   
35005                             sofa Bed And Breakfast     11   
46049            Big one bedroom apt, new and beautiful.     11   
31407                           Cozy feel at home studio     10   
33225                             Voted #1 Airbnb In NYC     10   
22261  Newly renovated, fully furnished room in Brooklyn     10   
22287                                            Jen Apt     10   
27972      Spacious 2-bedroom Apt in Heart of Greenpoint     10   
22835                            IT'S SIMPLY CONVENIENT!     10   
21700               Couch in Harlem Harvey Refugees only     10   
34446                                        Bronx Apart     10   
35601       Charming, bright and brand new Bed-Stuy home     10   
31066    Very Spacious bedroom, steps from CENTRAL PARK.     10   
47218                         Beautiful room in Bushwick     10   
35386                 Cozy room in a three-bedroom House     10   
33505                                   Room with a view     10   
2860       Large furnished 2 bedrooms- - 30 days Minimum     10   
23256             Quiet, Cozy UES Studio Near the Subway     10   
32810  Gigantic Sunny Room in Park Slope-Private Back...     10   
24100  Girls only, cozy room one block from Times Square     10   
25753                    Sunny, Quiet Room in Greenpoint      0   
26259                              the best you can find      0   
26866             Best Coliving space ever! Shared room.      0   
25634    MARTIAL LOFT 3: REDEMPTION (upstairs, 2nd room)      0   
23161  Huge Brooklyn Brownstone Living, Close to it all.      0   
25794  Spacious comfortable master bedroom with nice ...      0   
25778      Modern apartment in the heart of Williamsburg      0   
25796       Cozy yet spacious private brownstone bedroom      0   
25433      ★Hostel Style Room | Ideal Traveling Buddies★      0   
26841  Coliving in Brooklyn! Mod

Previously we showed the 99th percentile for proce to be around $800 a night. The listings above far exceed that, some by more than 10x. From the descriptions, these seem to be luxury spaces, event spaces, longer term leases, or mistakes. A google search showed rooms at the 5 star Four Seasons hotel to be going for around $1100, which helps make the case that these rooms are above and beyond the usual accommodations. These would likely not be of interest to your average tourist, we may consider excluding listings above $800 a night based on the 99th percentile. We might consider the same treatment on the low end, where some rooms are listed for $0. Again these appear to actual listings but likely have special circumstances that would not be interesting to your average tourist. We can also search for a way to exclude lease style listings.

In [30]:
airbnb[['name','price','minimum_nights','room_type','neighbourhood']].sort_values(by='minimum_nights', ascending=False).head(20)

name  price  \
5767                    Prime W. Village location 1 bdrm    180   
2854                                                 NaN    400   
38664                       Shared Studio (females only)    110   
13404                 Historic Designer 2 Bed. Apartment     99   
26341                    Beautiful place in Brooklyn! #2     79   
47620                             Williamsburg Apartment    140   
14285                    Peaceful apartment close to F/G     45   
8014                           Wonderful Large 1 bedroom     75   
11193                 Zen Room in Crown Heights Brooklyn     50   
7355                 Beautiful Fully Furnished 1 bed/bth    134   
10829                Brand New Luxury Apt Lease Takeover    199   
34487                                      Not available     50   
1305                 800sqft apartment with huge terrace    115   
15946                 Bright Luxury Studio, Central Loc.    250   
48446     Beautiful home on Staten Island, walk to ferry    100   
2831                    Luxury Gramercy Lg 1Bd w Balcony    250   
17338  Luxury drmn Bldg + Empire State Views & Roof Top!    189   
44429                 Room with sofa bed or air mattress   2000   
1449                  Manhattan Sky Crib (1 year sublet)    135   
20096  THANKSGIVING up to 6 GUESTS + HOT TUB + FIREPLACE    225   

       minimum_nights        room_type       neighbourhood  
5767             1250  Entire home/apt   Greenwich Village  
2854             1000  Entire home/apt   Battery Park City  
38664             999      Shared room   Greenwich Village  
13404             999  Entire home/apt              Harlem  
26341             999     Private room        Williamsburg  
47620             500  Entire home/apt        Williamsburg  
14285             500     Private room          Kensington  
8014              500  Entire home/apt              Harlem  
11193             500     Private room       Crown Heights  
7355              500  Entire home/apt    Long Island City  
10829             480  Entire home/apt    Long Island City  
34487             400     Private room  Bedford-Stuyvesant  
1305              370  Entire home/apt            Bushwick  
15946             366  Entire home/apt             Midtown  
48446             365  Entire home/apt          St. George  
2831              365  Entire home/apt            Gramercy  
17338             365  Entire home/apt            Kips Bay  
44429             365     Private room             Astoria  
1449              365  Entire home/apt             Chelsea  
20096             365     Private room        Clinton Hill

Looking at minimum_nights, the top few observations appear to be mistakes. Apartments often lease for 1 year, so values around 365 seem reasonable. 500 seams odd, and 999 and above just seems wrong. We could probably make the case for removal of those observations above 500, and potentially above 365 as well. The data discarded would be small compared to the overall data set. 

In [34]:
airbnb[['name','price','minimum_nights','number_of_reviews','room_type','neighbourhood','availability_365']].sort_values(by='number_of_reviews', ascending=False).head(20)

name  price  \
11759                            Room near JFK Queen Bed     47   
2031                          Great Bedroom in Manhattan     49   
2030                      Beautiful Bedroom in Manhattan     49   
2015                        Private Bedroom in Manhattan     49   
13495                            Room Near JFK Twin Beds     47   
10623                  Steps away from Laguardia airport     46   
1879            Manhattan Lux Loft.Like.Love.Lots.Look !     99   
20403  Cozy Room Family Home LGA Airport NO CLEANING FEE     48   
4870                  Private brownstone studio Brooklyn    160   
471                      LG Private Room/Family Friendly     60   
1271                    My Little Guest Room in Flushing     55   
113                         Park Slope Green Guest House    120   
8001                     Bright LARGE BED near Manhattan     65   
13612                Only Steps away from LaGuardia arpt     45   
225                  PRIVATE Room on Historic Sugar Hill     60   
1103                                    yahmanscrashpads     39   
1190                    Cozy Room in Lively East Village     72   
9323              Room steps away from LaGuardia airport     45   
2163                 TriBeCa 2500 Sq Ft w/ Priv Elevator    575   
257    Safe cute near subway& Manhattan NY NY retro s...     99   

       minimum_nights  number_of_reviews        room_type    neighbourhood  \
11759               1                629     Private room          Jamaica   
2031                1                607     Private room           Harlem   
2030                1                597     Private room           Harlem   
2015                1                594     Private room           Harlem   
13495               1                576     Private room          Jamaica   
10623               1                543     Private room    East Elmhurst   
1879                2                540     Private room  Lower East Side   
20403               1                510     Private room    East Elmhurst   
4870                1                488  Entire home/apt       Park Slope   
471                 3                480     Private room         Bushwick   
1271                1                474     Private room         Flushing   
113                30                467  Entire home/apt      South Slope   
8001                1                466     Private room    East Elmhurst   
13612               1                459     Private room    East Elmhurst   
225                 1                458     Private room           Harlem   
1103                1                454      Shared room          Jamaica   
1190                5                451     Private room     East Village   
9323                1                448     Private room    East Elmhurst   
2163                1                447  Entire home/apt          Tribeca   
257                 5                441  Entire home/apt          Astoria   

       availability_365  
11759               333  
2031                293  
2030                342  
2015                339  
13495               173  
10623               163  
1879                179  
20403               341  
4870                269  
471                   0  
1271                332  
113                 192  
8001                331  
13612               175  
225                 258  
1103                353  
1190                 13  
9323                166  
2163                207  
257                 226

Looking at number of reviews, these values are very high but at first glance I have no reason to suspect they are not real. It would take time to accumulate this many reviews, but I have no date information to give me clues that they are invalid. Depending on the algorithm used, we could keep them or possibly bin them to help with them potentially over influencing the model

In [36]:
airbnb[['name','price','reviews_per_month','number_of_reviews','minimum_nights','room_type','neighbourhood','availability_365']].sort_values(by='reviews_per_month', ascending=False).head(20)

name  price  \
42075  Enjoy great views of the City in our Deluxe Room!    100   
42076           Great Room in the heart of Times Square!    199   
38870                 Lou's Palace-So much for so little     45   
27287  JFK Comfort.5 Mins from JFK Private Bedroom & ...     80   
28651      JFK 2 Comfort 5 Mins from JFK Private Bedroom     50   
29628      JFK 3 Comfort 5 Mins from JFK Private Bedroom     50   
20403  Cozy Room Family Home LGA Airport NO CLEANING FEE     48   
22469  Cute Tiny Room Family Home by LGA NO CLEANING FEE     48   
36238                                 “For Heaven Cakes”     75   
40297        Studio Apartment 6 minutes from JFK Airport     67   
40036            6 Minutes From JFK Airport Cozy Bedroom     30   
42077   Nice Room 1 block away from Times Square action!    100   
11759                            Room near JFK Queen Bed     47   
21517   Sun Room Family Home LGA Airport NO CLEANING FEE     48   
44764  Hello! This is a very cozy space in Williamsburg.    100   
46937  ★ AMAZING★  TIME SQUARE/ 2 Bedroom 3 Bed Apart...    500   
21879                Private bedroom 8 mins-JFK&The Mall     65   
7086   Comfy Room Family Home LGA Airport NO CLEANING...     54   
46057           Balcony with Room! Close to LGA and JFK.     65   
13495                            Room Near JFK Twin Beds     47   

       reviews_per_month  number_of_reviews  minimum_nights        room_type  \
42075              58.50                156               1     Private room   
42076              27.95                 82               1     Private room   
38870              20.94                 37               1     Private room   
27287              19.75                403               1     Private room   
28651              17.82                341               1     Private room   
29628              16.81                302               1     Private room   
20403              16.22                510               1     Private room   
22469              16.03                436               1     Private room   
36238              15.78                132               1  Entire home/apt   
40297              15.32                 95               1     Private room   
40036              15.23                 98               1     Private room   
42077              14.62                 38               1     Private room   
11759              14.58                629               1     Private room   
21517              14.36                417               1     Private room   
44764              14.00                 28               1     Private room   
46937              14.00                 14               3  Entire home/apt   
21879              13.48                347               1     Private room   
7086               13.45                430               1     Private room   
46057              13.42                 17               1     Private room   
13495              13.40                576               1     Private room   

             neighbourhood  availability_365  
42075     Theater District               299  
42076     Theater District               299  
38870             Rosedale               134  
27287  Springfield Gardens                26  
28651  Springfield Gardens                25  
29628  Springfield Gardens                26  
20403        East Elmhurst               341  
22469        East Elmhurst               337  
36238  Springfield Gardens                28  
40297              Jamaica               145  
40036              Jamaica               189  
42077     Theater District               295  
11759              Jamaica               333  
21517        East Elmhurst               338  
44764         Williamsburg                20  
46937       Hell's Kitchen                45  
21879              Jamaica               159  
7086         East Elmhurst               347  
46057        Richmond Hill               157  
13495              Jama

For reviews per month, the top observation, 58.5, must be a mistake. The minimum stay is one night, so the maximum value should be 31. 